File: Rust/src/logger.rs
use std::fs::{File, OpenOptions};
use std::io::{Write, Result as IoResult};
use std::sync::mpsc::{Receiver, Sender};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{SystemTime, UNIX_EPOCH};
use chrono::Local;

// Logging configuration
pub struct Logger {
    log_file: Arc<Mutex<File>>,
}

impl Logger {
    // Create a new logger instance
    pub fn new() -> IoResult<Self> {
        let log_file = OpenOptions::new()
            .create(true)
            .write(true)
            .append(true)
            .open("./logs/kv_store_operations.log")?;

        Ok(Logger {
            log_file: Arc::new(Mutex::new(log_file)),
        })
    }

    // Log an operation
    pub fn log_operation(&self, operation_type: &str, key: &str, result: &str) -> IoResult<()> {
        let timestamp = Local::now().format("%a %b %e %T %Y").to_string();
        let log_entry = format!(
            "{} - {} - key: {}, result: {}\n", 
            timestamp, 
            operation_type.to_uppercase(), 
            key, 
            result
        );

        let mut file = self.log_file.lock().unwrap();
        file.write_all(log_entry.as_bytes())?;
        file.flush()?;

        Ok(())
    }

    // Log the entire KV store contents periodically
    pub fn log_entire_store(rx: Receiver<()>) {
        let log_file_path = "./logs/kv_store_contents.log";
        
        loop {
            // Sleep for 10 seconds
            thread::sleep(std::time::Duration::from_secs(10));

            // Check if shutdown signal received
            if rx.try_recv().is_ok() {
                break;
            }

            // Try to log store contents
            match OpenOptions::new()
                .create(true)
                .write(true)
                .truncate(true)
                .open(log_file_path) {
                Ok(mut file) => {
                    let timestamp = Local::now().to_string();
                    
                    // TODO: Replace with actual KV store retrieval
                    let store_contents = "Placeholder for KV store contents";
                    
                    if let Err(e) = writeln!(file, "{}: {}", timestamp, store_contents) {
                        eprintln!("Error writing to log file: {}", e);
                    }
                },
                Err(e) => {
                    eprintln!("Error opening log file: {}", e);
                }
            }
        }
    }
}

// Wrapper for global logging operations
pub fn log_operation(operation_type: &str, key: &str, result: &str) {
    // Note: In a real application, you'd use a global logger or pass a logger instance
    if let Ok(logger) = Logger::new() {
        let _ = logger.log_operation(operation_type, key, result);
    }
}

File: kv_store/src/logger.py
import logging
import threading
import time
from queue import Queue
from kv_store import kv_store

# Set up a logging queue
log_queue = Queue()

# Configure logger to use the queue handler
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("kv_store_operations")

# Log worker continuously writes from queue to logfile
def log_worker():
    with open("../logs/kv_store_operations.log", "a") as logfile:
        while True:
            log_entry = log_queue.get()
            if log_entry is None:
                break  # Exit signal received
            logfile.write(log_entry + "\n")
            logfile.flush()  # Ensure log is written to file immediately
    print("Log worker exiting")

# Start the log worker thread
log_thread = threading.Thread(target=log_worker, daemon=True)
log_thread.start()

# Get time & format log operation
def log_operation(operation_type, key, result):
    timestamp = time.ctime()
    log_entry = f"{timestamp} - {operation_type.upper()} - key: {key}, result: {result}"
    log_queue.put(log_entry)


# Periodic logging of entire kv_store
def log_entire_store():
    threading.Timer(10.0, log_entire_store).start()
    try:
        with open("./logs/kv_store_contents.log", "w") as logfile:
            logfile.write(f"{time.ctime()}: {kv_store.get('*')}\n")
    except Exception as e:
        print(f"Error writing to log file: {e}")
        
